ray-on-gke/user/kuberay-tpu-webhook/main.go

```diff
@@ -0,0 +1,266 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"strconv"
+
+	ray "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+)
+
+// our representation of a pod slice
+// not necessarily true that worker group scheduled on 1 slice
+type Slice struct {
+	rayClusterName	string
+	groupName	string
+}
+
+var (
+	tpuResourceName = corev1.ResourceName("google.com/tpu")
+	certPath = "/etc/kuberay-tpu-webhook/tls/tls.crt"
+	keyPath = "/etc/kuberay-tpu-webhook/tls/tls.key"
+
+	// map of ray cluster names to # of workers created in the slice
+	sliceToWorkers map[Slice]int
+)
+
+// check if containers are requesting TPU resources
+func containerRequestingTPUs(containers ...corev1.Container) bool {
+	for _, container := range containers {
+		if l := container.Resources.Limits; l != nil {
+			if resource := l[tpuResourceName]; !resource.IsZero() {
+				return true
+			}
+		}
+		if r := container.Resources.Requests; r != nil {
+			if resource := r[tpuResourceName]; !resource.IsZero() {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// check if request is for TPU multi-host
+func isTPUMultiHost(topology string) bool {
+	topologyVals :=  strings.Split(topology, "x")
+	values := make([]int, len(topologyVals))
+	for i := 0; i < len(values); i++ {
+		values[i], _ = strconv.Atoi(topologyVals[i])
```

Comment:
I believe it intercepts requests after GCW, since from what I can see the topology string is set correctly. I added error handling in the case that it intercepts before.